### PR TITLE
Update Readme and Contributing to be more consistent

### PR DESCRIPTION
**Pull request by ashthespy**
Pulled out from #352,  to pave the way for #385. 
We should probably document what the preferred contributing process is now.  

Avoid multiple copies of dependencies
Indicate that `rodio` is default backend
Remove `nightly` requirement for development
Fix minor typos

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -2,38 +2,30 @@
 
 ## Setup
 
-In order to contribute to librespot, you will first need to set up a suitable rust build environment, with the necessary dependenices installed. These instructions will walk you through setting up a simple build environment.
+In order to contribute to librespot, you will first need to set up a suitable Rust build environment, with the necessary dependencies installed. These instructions will walk you through setting up a simple build environment.
 
-You will need to have C compiler, rust, and portaudio installed.
+You will need to have C compiler, Rust, and alsa/portaudio libraries installed.
```
Comment (kingosticks) [inline CONTRIBUTING.md]: s/have C compiler/have a C compiler/


It's also better to say something like "and the development libraries for the audio backend(s) you want". You don't need any ALSA/PortAudio libraries if you are going with the default Rodio backend.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -2,38 +2,30 @@
...
-```bash
-sudo apt-get install build-essential portaudio19-dev
-```
-
-On Fedora systems, the following command will install these dependencies :
+Follow any prompts it gives you to install Rust. Once that’s done, Rust's standard tools should be setup and ready to use.
 
+#### Additional Rust tools - `rustfmt`
+To ensure a consistent codebase, install [`rustfmt`](https://github.com/rust-lang/rustfmt) via `rustup`.
+This is not optional, as Travis CI is set up to check that code is compliant with this repos style guides.
```
Comment (kingosticks) [inline CONTRIBUTING.md]: s/repos/repo's/ or s/this repos/our/.

Technically they are not style 'guides' if they are mandatory. Maybe 'rules' is more accurate.

File: README.md [hunk 3]
```diff
@@ -30,9 +30,11 @@ If you run into a bug when using librespot, please search the existing issues be
...
 
-We recently switched to using [Rodio](https://github.com/tomaka/rodio) for audio playback by default, hence for macOS and Windows, you should just be able to clone and build librespot (with the command below). For linux, you will need to run the additional commands below, depending on your distro.
+## Additional Dependencies
+We recently switched to using [Rodio](https://github.com/tomaka/rodio) for audio playback by default, hence for macOS and Windows, you should just be able to clone and build librespot (with the command below).
+For Linux, you will need to run the additional commands below, depending on your distro.
 
-On debian / ubuntu, the following command will install these dependencies :
+On Debian/Ubuntu, the following command will install these dependencies :
 ```shell
 sudo apt-get install build-essential libasound2-dev
```
Comment (kingosticks) [inline README.md]: Isn't `libasound2-dev` only required for the ALSA backend?

Comment (ashthespy) [reply README.md]: Rodio(Cpal) [as well](https://github.com/RustAudio/cpal/blob/8f28048f611dacb7d6ef2df4f26d951d6511f670/.github/workflows/cpal.yml#L12)?

File: CONTRIBUTING.md (hunk 1 above)
Comment (kingosticks) [reply CONTRIBUTING.md]: > Rodio uses Cpal which in turn uses Alsa/Winapi, so will still need ALSA right?

I've never tried to build the Rodio backend but I assume not. Sounds like that assumption was total rubbish. In which case:

s/alsa/ALSA/
s/PortAudio/

**Comment by sashahilton00** [comment]
Sounds like a good idea.

**Comment by willstott101** [comment]
I think compiling isn't just something contributors would do. It feels wrong to me to have most of this information in CONTRIBUTING

Maybe it would be better to have a COMPILING in addition to CONTRIBUTING, where the later is more focused on workflow/etiquette, and the former on build environment stuff.

Github wikis always feel a bit weird to me, so I'm not against moving stuff into the repo.

**Comment by ashthespy** [comment]
>Github wikis always feel a bit weird to me, so I'm not against moving stuff into the repo.

Same, it feels weird not being able to document additional feature-flags/option-flags in the commits/PR that introduces them.

How about short and succulent README (as we have now), and the nitty-gritty details of compilation broken into it own page. I will split the details in CONTRIBUTING out into COMPILING then.

**Comment by sashahilton00** [comment]
I'm ok with splitting into files, might I suggest we have the `CONTRIBUTING.md` file dedicated to requirements for contributing and such, whilst we just create a `WIKI.md` for the other stuff that currently resides in the wiki? I agree that the Github wiki system is pretty clunky, I don't know why they don't just allow people to designate a `_wiki` folder for wiki files. Either way, what I think we should avoid is a scenario where some of the info currently on the wiki ends up in markdown files, other stuff in the existing wiki. Personally I'd be in favour of moving it all to markdown files and disabling the wiki entirely.

**Comment by ashthespy** [comment]
I broke out all the compilation parts into `COMPILING.md` - give it a quick read and see if it's fine? 
Next would be to decide on what should go into `CONTRIBUTING.md` - but will leave that to the rest to decide about `rustfmt` and the likes and a different PR.

File: COMPILING.md [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+```
+On Fedora systems, install with:
+```shell
+sudo dnf install gcc 
+```
+### Audio library dependencies
+Depending on the chosen backend, specific development libraries are required. 
+
+*_Note this is an non extensive list, open a PR to add to it!_*
```
Comment (kingosticks) [inline COMPILING.md]: Do you mean 'a non-exhaustive list'?

File: COMPILING.md [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+cd librespot
+```
+
+## Compiling & Running
+
+Once your build environment is setup, compiling the code is pretty simple.
+
+### Compiling
+
+To build a ```debug``` build, from the project root:
```
Comment (kingosticks) [inline COMPILING.md]: I think you should explicitly mention this command will build with the default backend as it section just above explicitly mentions building for the ALSA backend.

File: COMPILING.md [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+cargo build
+```
+
+And for ```release```:
+
+```bash
+cargo build --release
+```
+
+You will most likely want to build debug builds when developing, as they are faster, and more verbose, for the purposes of debugging.
```
Comment (kingosticks) [inline COMPILING.md]: I'd reword the benefits of the debug build. I can see what you are saying but it could easily be misunderstood that the debug build *runs* faster. You should maybe say that a debug build is required/useful when submitting a bug report.

File: COMPILING.md [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+There are also a number of compiler feature flags that you can add, in the event that you want to have certain additional features also compiled. The list of these is available on the [wiki](https://github.com/librespot-org/librespot/wiki/Compiling#addition-features).
+
+By default, librespot compiles with the ```rodio-backend``` feature. To compile without default features, you can run with:
+
+```bash
+cargo build --no-default-features
+```
+
```
Comment (kingosticks) [inline COMPILING.md]: I'd give an example of how to compile with only alsa (or whatever backend) here too.

File: COMPILING.md [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+## Setup
+
+In order to compile librespot, you will first need to set up a suitable Rust build environment, with the necessary dependencies installed. You will need to have a C compiler, Rust, and the development libraries for the audio backend(s) you want installed. These instructions will walk you through setting up a simple build environment.
+
+### Install Rust
+The easiest, and recommended way to get Rust is to use [rustup](https://rustup.rs). You can install `rustup` with this command:
+
+```bash
+curl https://sh.rustup.rs -sSf | sh
```
Comment (willstott101) [inline COMPILING.md]: Perhaps we only need the link to the rustup website, as obviously this command doesn't work on windows. Their website has a nice landing page with specific instructions for a detected OS.

File: README.md [hunk 9]
```diff
@@ -36,9 +36,11 @@ If you run into a bug when using librespot, please search the existing issues be
 # Building
 Rust 1.32.0 or later is required to build librespot.
```
Comment (willstott101) [inline README.md]: Should we just link to COMPILING.md for all of this section now?

File: COMPILING.md [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+```bash
+cargo build --no-default-features --features "alsa-backend"
+```
+
+### Running
+
+Assuming you just compiled a ```debug``` build, you can run librespot with the following command:
+
+```bash
+./target/debug/librespot -n Librespot
```
Comment (willstott101) [inline COMPILING.md]: It would be nice to mention `cargo run -- -n Librespot` as a minimal quick-start around here. I find that's a much nicer way to work.

File: README.md [hunk 11]
```diff
@@ -48,7 +50,19 @@ On Fedora systems, the following command will install these dependencies :
...
+```
+Rodio (default)
+ALSA
+PortAudio
+PulseAudio
+JACK
+SDL
+Pipe
+```
+Please check the corresponding [compiling wiki entry](https://github.com/librespot-org/librespot/wiki/Compiling#dependencies) for backend specific dependencies.
```
Comment (willstott101) [inline README.md]: Or at the very least this should link to COMPILING. I _think_ relative links are supported.

File: README.md (hunk 9 above)
Comment (sashahilton00) [reply README.md]: Probably. We should also change the `Rust x.x.x or later is required`. To something like `The minimum version of rust required can be found here.` and link through to the minimum version specified in the travis.yml file.

File: COMPILING.md (hunk 10 above)
Comment (sashahilton00) [reply COMPILING.md]: depends. the current way makes it clear where the compiled binary is located.

Comment (willstott101) [reply COMPILING.md]: That's true, and if we're not careful we'll just end up duplicating the cargo docs here :)

I'm happy with it as is

File: README.md (hunk 9 above)
Comment (ashthespy) [reply README.md]: Sorry, I didn't update the Readme, will do it this evening. 
> 
> 
> Probably. We should also change the `Rust x.x.x or later is required`. To something like `The minimum version of rust required can be found here.` and link through to the minimum version specified in the travis.yml file.

I would rather people bump the readme when they bump Travis, rather than the other way around. Also if/when we move to Github actions/CircleCI or any other CI, it becomes another broken link if we aren't careful.

File: README.md (hunk 11 above)
Comment (ashthespy) [reply README.md]: Fixed.

**Comment by ashthespy** [comment]
> 
> 
> Looking good so far! The only part I see missing is https://github.com/librespot-org/librespot/wiki/Compile-librespot-for-kernel-prior-3.9 but the effort is probably better spent with getting [plietar/rust-mdns@66a7403](https://github.com/plietar/rust-mdns/commit/66a74033da6c9f1a06e7b0a29f4544fd189d6479) into libmdns.

Left that in the wiki as its quite specific, should be fine right?

**Comment by willstott101** [comment]
Yeah I think that'll be alright, I've never come across the issue myself but afaik it's fixed in https://github.com/rampantmonkey/libmdns/pull/7

**Comment by ashthespy** [comment]
@willstott101 neat, lets hope it gets merged, that repo doesn't seem extremely active. 

Any objections merging this in the meantime?

**Comment by sashahilton00** [comment]
nope feel free to merge.
